blaze-client/src/test/scala/org/http4s/blaze/client/BlazeClientConnectionReuseSuite.scala

```diff
@@ -0,0 +1,197 @@
+  }
+
+  test(
+    "BlazeClient.status shouldn't wait for the response entity and shouldn't reuse the connection if the response entity wasn't fully read") {
+    val servers = makeServers()
+    builder().resourceWithState.use { case (client, state) =>
+      for {
+        _ <- client
+          .status(Request[IO](GET, servers(0) / "huge"))
+          .timeout(5.seconds) // we expect it to complete without waiting for the response body
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "simple"))
+        _ <- state.totalAllocations.assertEquals(2L)
+      } yield ()
+    }
+  }
+
+  test("BlazeClient should reuse connection to different servers separately".flaky) {
+    val servers = makeServers()
+    builder().resourceWithState.use { case (client, state) =>
+      for {
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "simple"))
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "simple"))
+        _ <- state.totalAllocations.assertEquals(1L)
+        _ <- client.expect[String](Request[IO](GET, servers(1) / "simple"))
+        _ <- client.expect[String](Request[IO](GET, servers(1) / "simple"))
+        _ <- state.totalAllocations.assertEquals(2L)
+      } yield ()
+    }
+  }
+
+  //// Decoding failures ////
+
+  test("BlazeClient should reuse the connection after response decoding failed".flaky) {
+    // This will work regardless of whether we drain the entity or not,
+    // because the response is small and it is read in full in first read operation
+    val servers = makeServers()
+    val drainThenFail = EntityDecoder.error[IO, String](new Exception())
+    builder().resourceWithState.use { case (client, state) =>
+      for {
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "simple"))(drainThenFail).attempt
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "simple"))
+        _ <- state.totalAllocations.assertEquals(1L)
+      } yield ()
+    }
+  }
+
+  test(
+    "BlazeClient should reuse the connection after response decoding failed and the (large) entity was drained".fail) {
+    val servers = makeServers()
+    val drainThenFail = EntityDecoder.error[IO, String](new Exception())
+    builder().resourceWithState.use { case (client, state) =>
+      for {
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "large"))(drainThenFail).attempt
+        _ <- client.expect[String](Request[IO](GET, servers(0) / "simple"))
+        _ <- state.totalAllocations.assertEquals(1L)
+      } yield ()
+    }
+  }
+
+  test(
+    "BlazeClient shouldn't reuse the connection after response decoding failed and the (large) entity wasn't drained") {
+    val servers = makeServers()
+    val failWithoutDraining = new EntityDecoder[IO, String] {
+      override def decode(m: Media[IO], strict: Boolean): DecodeResult[IO, String] =
+        DecodeResult[IO, String](IO.raiseError(new Exception()))
+      override def consumes: Set[MediaRange] = Set.empty
+    }
```

Comment:
I would not have passed a quiz on how this differs from `EntityDecoder.error` if I weren't reading this!